package_test.go

```diff
@@ -0,0 +1,166 @@
+
+//go:embed tests/test_runner.sh
+var testRunnerCmd string
+
+//go:embed tests/test.sh
+var testSH string
+
+func testPackage(ctx context.Context, t *testing.T, client *dagger.Client, spec *archive.Spec) {
+	// set up the daemon container
+	getContainer, ok := distros[spec.Distro]
+	if !ok {
+		t.Fatalf("unknown distro: %s", spec.Distro)
+	}
+
+	buildOutput, err := do(ctx, client, spec)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	batsCore, batsHelpers := makeBats(client)
+
+	qemu := testutil.NewQemuImg(ctx, client)
+
+	c := getContainer(ctx, t, client).
+		WithDirectory("/opt/bats", batsCore).
+		WithExec([]string{"/bin/sh", "-c", "cd /opt/bats && ./install.sh /usr/local"}).
+		WithDirectory("/opt/moby/test_helper", batsHelpers).
+		WithNewFile("/opt/moby/test.sh", dagger.ContainerWithNewFileOpts{Contents: testSH, Permissions: 0744}).
+		WithDirectory("/lib/modules", qemu.Directory("/lib/modules"))
+
+	goCtr, err := targets.InstallGo(ctx, c, client.CacheVolume(targets.GoModCacheKey), client.CacheVolume("jammy-go-build-cache-"+spec.Arch))
+	if err != nil {
+		t.Fatal(err)
+	}
+	goCtr = goCtr.WithEnvVariable("CGO_ENABLED", "0")
+
+	aptly := goCtr.WithExec([]string{"go", "install", "github.com/aptly-dev/aptly@v1.5.0"}).File("/go/bin/aptly")
+	c = c.WithFile("/usr/local/bin/aptly", aptly)
+
+	entrypointBin := goCtr.
+		WithExec([]string{"go", "install", "github.com/cpuguy83/qemu-micro-env/cmd/entrypoint@" + entrypointVersion}).
+		File("/go/bin/entrypoint")
+
+	resolvConf, err := c.WithExec([]string{"cat", "/etc/resolv.conf"}).Stdout(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	rootfs := c.WithNewFile("/usr/local/bin/setup_ssh", dagger.ContainerWithNewFileOpts{
+		Contents:    setupSSH,
+		Permissions: 0744,
+	}).
+		WithNewFile("/lib/systemd/system/setup_ssh.service", dagger.ContainerWithNewFileOpts{
+			Contents:    setupSSHService,
+			Permissions: 0644,
+		}).
+		WithExec([]string{"systemctl", "enable", "setup_ssh.service"}).Rootfs().
+		WithNewFile("/etc/resolv.conf", resolvConf)
+
+	qcow := testutil.QcowFromDir(ctx, rootfs, qemu)
+
+	buf := make([]byte, 16)
+	n, err := rand.Read(buf)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	sockets := client.CacheVolume("qemu-micro-env-sockets-" + hex.EncodeToString(buf[:n]))
+
+	runner := qemu.
+		WithMountedFile("/tmp/rootfs-base.qcow2", qcow).
+		WithMountedFile("/usr/local/bin/docker-entrypoint", entrypointBin).
+		WithMountedCache("/tmp/sockets", sockets).
+		WithNewFile("/usr/local/bin/docker-entrypoint.sh", dagger.ContainerWithNewFileOpts{Contents: entrypointCmd, Permissions: 0744}).
+		WithExec([]string{"/bin/sh", "-c", "chown -R 65534:65534 /tmp/sockets"}).
+		WithEnvVariable("DEBUG", strconv.FormatBool(flDebug)).
+		WithExec([]string{"docker-entrypoint.sh"}, dagger.ContainerWithExecOpts{
+			InsecureRootCapabilities: true,
+		}).
+		WithExposedPort(22, dagger.ContainerWithExposedPortOpts{Protocol: dagger.Tcp, Description: "VM ssh"})
+
+	const svc = "testvm"
+
+	testRunner := qemu.
```

Comment:
How about renaming these to `runnerServer` and `runnerClient`? Took me a minute to figure out what was going on.